File: Components/Strata.Foundation.Core/main/java/strata/foundation/core/utility/Consumed.java
//////////////////////////////////////////////////////////////////////////////
// Consumed.java
//////////////////////////////////////////////////////////////////////////////

package strata.foundation.core.utility;

import java.util.Objects;
import java.util.Optional;

/**
 * <p>
 * Result produced by an operation that accepts an input but returns
 * no output, capturing the method name, the input value, and, when
 * the invocation failed, the resulting {@link Throwable}. This
 * covers any consume-only scenario — {@link java.util.function.Consumer}
 * invocations, event dispatchers, loggers, validators that throw on
 * failure, or any void method that receives a single argument.
 * </p><br/>
 * <b>Examples</b><br/>
 * <pre>
 * Consumed&lt;String&gt; result = Consumed.of("process",input);
 *
 * if (result.isFailure())
 *     log(result.getException());
 * </pre>
 *
 * @param <I> - the type of the input consumed by the operation
 */
public
class Consumed<I>
{
    private final String    method;
    private final I         input;
    private       Throwable throwable;

    public
    Consumed(String method,I input)
    {
        this.method    = Objects.requireNonNull(method,"method");
        this.input     = Objects.requireNonNull(input,"input");
        this.throwable = null;
    }

    public
    Consumed(String method,I input,Throwable throwable)
    {
        this.method = Objects.requireNonNull(method,"method");
        this.input  = Objects.requireNonNull(input,"input");

        try
        {
            this.throwable =
                Objects.requireNonNull(throwable,"throwable");
        }
        catch (Exception e)
        {
            this.throwable = e;
        }
    }

    public String
    getMethod() { return method; }

    public I
    getInput() { return input; }

    public Optional<Throwable>
    getException() { return Optional.ofNullable(throwable); }

    public boolean
    isSuccess() { return this.throwable == null; }

    public boolean
    isFailure() { return this.throwable != null; }

    public static <I> Consumed<I>
    of(String method,I input) { return new Consumed<I>(method,input); }

    public static <I> Consumed<I>
    of(String method,I input,Throwable throwable) { return new Consumed<I>(method,input,throwable); }
}

//////////////////////////////////////////////////////////////////////////////

File: Components/Strata.Foundation.Core/main/typescript/Utility/Consumed.ts
//////////////////////////////////////////////////////////////////////////////
// Consumed.ts
//////////////////////////////////////////////////////////////////////////////

import {Optional}              from "./Optional";
import {NullPointerException}  from "./NullPointerException";

export
class Consumed<I>
{
    private readonly method:    string;
    private readonly input:     I;
    private readonly throwable: Error | null;

    private
    constructor(method: string,input: I,throwable: Error | null)
    {
        this.method    = method;
        this.input     = input;
        this.throwable = throwable;
    }

    getMethod(): string { return this.method; }

    getInput(): I { return this.input; }

    getException(): Optional<Error> { return Optional.ofNullable<Error>(this.throwable); }

    isSuccess(): boolean { return this.throwable == null; }

    isFailure(): boolean { return this.throwable != null; }

    static of<I>(method: string,input: I): Consumed<I>;
    static of<I>(method: string,input: I,throwable: Error): Consumed<I>;
    static of<I>(method: string,input: I,throwable?: Error): Consumed<I>
    {
        if (method == null)
            throw new NullPointerException("method is null");

        if (input == null)
            throw new NullPointerException("input is null");

        if (throwable !== undefined)
        {
            if (throwable == null)
                return new Consumed<I>(method,input,new NullPointerException("throwable is null"));

            return new Consumed<I>(method,input,throwable);
        }

        return new Consumed<I>(method,input,null);
    }
}

//////////////////////////////////////////////////////////////////////////////
